<a href="https://colab.research.google.com/github/jsansao/teic_tp_mock/blob/main/tp1_penguins.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Exemplo de Trabalho Prático 1 (TP1) - Ciência de Dados
## Dataset: Palmer Penguins (Classificação de Espécies de Pinguins)

Este notebook serve como modelo exemplar de resolução do Trabalho Prático 1 para a disciplina de Ciências de Dados. A estrutura apresentada atende a todos os requisitos do documento de orientações, demonstrando o rigor científico, a formalização metodológica e o desenvolvimento de um protótipo de baseline usando AutoML.

---

## 1. Formalização do Problema

### 1.1 Contextualização, Motivação e Importância
O dataset *Palmer Penguins* foi coletado e disponibilizado pela Dra. Kristen Gorman e a Estação de Pesquisa Antártica LTER Palmer. Ele registra medições biológicas e dados de amostragem ecológica para três espécies de pinguins (*Adelie*, *Chinstrap* e *Gentoo*) observadas no Arquipélago Palmer, na Antártica.

A motivação deste estudo reside na classificação taxonômica automatizada a partir de medições morfológicas corporais não invasivas (como tamanho do bico, nadadeira e massa corporal). A importância dessa modelagem está no suporte a biólogos e ecologistas no monitoramento populacional de fauna antártica, otimizando a identificação em campo e reduzindo a necessidade de exames genéticos custosos.

### 1.2 Abordagem Analítica
Para investigar este problema, adotaremos uma abordagem de **Classificação Multiclasse Supervisionada**. O conjunto de dados possui 3 classes exclusivas correspondentes às espécies de pinguim ($Y \in \{\text{'Adelie'}, \text{'Chinstrap'}, \text{'Gentoo'}\}$). O objetivo é mapear um vetor de características morfológicas e ambientais na espécie correta.

---

## 2. Hipótese a ser Investigada

### 2.1 Hipótese
> **Hipótese de Trabalho ($H_1$):** As medições morfológicas corporais (comprimento e profundidade do bico, comprimento da nadadeira e massa corporal), combinadas com o local geográfico (ilha de ocorrência), apresentam variações estatísticas suficientemente distintas entre as espécies *Adelie*, *Chinstrap* e *Gentoo*, permitindo a classificação taxonômica com altíssima acurácia por algoritmos de aprendizado de máquina.

### 2.2 Objetivo Geral
Construir um modelo preditivo preliminar de baseline utilizando AutoML para classificar as três espécies do dataset *Palmer Penguins*, avaliando o desempenho inicial da solução.

### 2.3 Objetivos Específicos
1. Carregar o dataset *Palmer Penguins* de forma transparente via repositório/biblioteca pública.
2. Inspecionar as características estruturais da base (volume de observações, tipos de dados e presença de nulos).
3. Detalhar e classificar cada uma das variáveis preditoras e a variável alvo.
4. Treinar múltiplos algoritmos de classificação automática usando uma biblioteca de AutoML (FLAML), selecionando o modelo baseline.
5. Avaliar o desempenho em relação à estimativa trivial da moda (espécie mais frequente).

### 2.4 Metodologia Experimental
1. **Aquisição dos Dados:** Carregamento direto via `seaborn` ou repositório oficial no GitHub.
2. **Diagnóstico dos Dados:** Inspecionar dimensões (`shape`), primeiras linhas (`head`) e tipos de dados (`info`), sem realizar gráficos ou transformações manuais avançadas nesta etapa.
3. **Modelagem via AutoML:** Uso da biblioteca FLAML com limite rígido de tempo de 60 segundos para selecionar o melhor modelo.
4. **Avaliação:** Medição de Acurácia global e F1-Score Macro em validação cruzada/teste.

---

## 3. Revisão Bibliográfica

### 3.1 Referências de Artigos Científicos
1. **Gorman, K. B., Williams, T. D., & Nygard, W. R. (2014).** *Ecological Sexual Dimorphism and Environmental Variability within a Community of Antarctic Penguins (Gen. Pygoscelis).* PLOS ONE, 9(3), e90081.
   * **Relevância:** Artigo científico original que descreve a amostragem de campo e valida a diferenciação morfológica entre as três espécies no Arquipélago Palmer.
2. **Horst, A. M., Hill, A. P., & Gorman, K. B. (2020).** *palmerpenguins: Palmer Archipelago (Antarctica) penguin data.* R package version 0.1.0.
   * **Relevância:** Publicação que introduziu o conjunto de dados como uma alternativa moderna ao dataset *Iris* para ensino e benchmark de Ciência de Dados.

### 3.2 Projetos Semelhantes no GitHub
1. **Palmer Penguins Classification in Python**
   * Link: [https://github.com/allisonhorst/palmerpenguins](https://github.com/allisonhorst/palmerpenguins)
   * **Relevância:** Repositório oficial do dataset com exemplos de pipelines em R e Python.
2. **Scikit-Learn Classification Benchmarks on Penguins**
   * Link: [https://github.com/mcnulty/palmerpenguins](https://github.com/mcnulty/palmerpenguins)
   * **Relevância:** Demonstrações de pipelines de classificação tabular usando Scikit-Learn.

### 3.3 Integração no Contexto do Trabalho
Os estudos da Dra. Gorman comprovam que as medições morfológicas possuem alto valor discriminatório. Utilizaremos a ferramenta de AutoML para testar se algoritmos tabulares conseguem separar as três espécies automaticamente a partir desses dados.

---

## 4. Apresentação da Base de Dados

### 4.1 Por que essa base é adequada?
O dataset *Palmer Penguins* é ideal por ser um conjunto de dados real, equilibrado e representativo de problemas ecológicos e biológicos. Ele possui um excelente balanceamento de variáveis numéricas contínuas e variáveis categóricas ambientais.

### 4.2 Carregamento e Características Básicas

In [1]:
# Instalação da biblioteca FLAML no ambiente Google Colab
!pip install flaml -q

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 349.7/349.7 kB 3.3 MB/s eta 0:00:00


In [2]:
import pandas as pd
import numpy as np
import seaborn as sns

# Carregamento da base através do Seaborn ou URL pública confiável
url = "https://raw.githubusercontent.com/allisonhorst/palmerpenguins/master/inst/extdata/penguins.csv"
df = pd.read_csv(url)

print("Formato dos dados (linhas, colunas):", df.shape)
print("-" * 50)
df.head()

Formato dos dados (linhas, colunas): (344, 8)
--------------------------------------------------


,species,island,bill_length_mm,bill_depth_mm,flipper_length_mm,body_mass_g,sex,year
0,Adelie,Torgersen,39.1,18.7,181.0,3750.0,male,2007
1,Adelie,Torgersen,39.5,17.4,186.0,3800.0,female,2007
2,Adelie,Torgersen,40.3,18.0,195.0,3250.0,female,2007
3,Adelie,Torgersen,NaN,NaN,NaN,NaN,NaN,2007
4,Adelie,Torgersen,36.7,19.3,193.0,3450.0,female,2007


In [3]:
# Tipos de dados e verificação inicial de nulos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 344 entries, 0 to 343
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   species            344 non-null    object 
 1   island             344 non-null    object 
 2   bill_length_mm     342 non-null    float64
 3   bill_depth_mm      342 non-null    float64
 4   flipper_length_mm  342 non-null    float64
 5   body_mass_g        342 non-null    float64
 6   sex                333 non-null    object 
 7   year               344 non-null    int64  
dtypes: float64(4), int64(1), object(3)
memory usage: 21.6+ KB


### 4.3 Detalhamento das Variáveis (Features)

*   **species (Target):** Espécie do pinguim (`Adelie`, `Chinstrap`, `Gentoo`). Variável categórica nominal multiclasse (3 classes).
*   **island:** Ilha do Arquipélago Palmer onde a amostra foi coletada (`Biscoe`, `Dream`, `Torgersen`). Variável categórica nominal.
*   **bill_length_mm:** Comprimento do bico (culmen) em milímetros. Variável numérica contínua.
*   **bill_depth_mm:** Profundidade/altura do bico (culmen) em milímetros. Variável numérica contínua.
*   **flipper_length_mm:** Comprimento da nadadeira em milímetros. Variável numérica contínua.
*   **body_mass_g:** Massa corporal em gramas. Variável numérica contínua.
*   **sex:** Sexo do pinguim (`male`, `female`). Variável categórica nominal.
*   **year:** Ano da amostragem (2007, 2008, 2009). Variável numérica discreta.

---

## 5. AutoML: Prova de Conceito e Baseline

### 5.1 Justificativa de Escolha do FLAML
A biblioteca **FLAML** foi escolhida por ser leve, rápida e imune a conflitos de versão no Google Colab. Ela busca automaticamente os melhores classificadores tabulares (como LightGBM, XGBoost e Random Forest) de forma otimizada.

### 5.2 Métrica de Avaliação Escolhida
A métrica principal será a **Acurácia (Accuracy)** e o **F1-Score Macro**, adequados para avaliar a performance global nas 3 espécies.

### 5.3 Implementação do AutoML

In [4]:
from sklearn.model_selection import train_test_split
from flaml import AutoML

# Remoção de poucas linhas nulas para a Prova de Conceito básica do AutoML
df_clean = df.dropna().copy()

# Conversão de colunas categóricas para o tipo 'category' do pandas
df_clean['island'] = df_clean['island'].astype('category')
df_clean['sex'] = df_clean['sex'].astype('category')

X = df_clean.drop(columns=['species'])
y = df_clean['species']

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

# Inicialização do FLAML
automl = AutoML()
settings = {
    "time_budget": 60,            # 60 segundos de busca
    "metric": 'accuracy',
    "task": 'classification',
    "log_file_name": 'penguins_automl.log',
    "seed": 42,
    "verbose": 0
}

print("Executando AutoML para classificação de Pinguins (limite de 60s)...")
automl.fit(X_train=X_train, y_train=y_train, **settings)
print("Busca finalizada!")

Executando AutoML para classificação de Pinguins (limite de 60s)...


INFO:flaml.tune.searcher.blendsearch:No low-cost partial config given to the search algorithm. For cost-frugal search, consider providing low-cost values for cost-related hps via 'low_cost_partial_config'. More info can be found at https://microsoft.github.io/FLAML/docs/FAQ#about-low_cost_partial_config-in-tune
INFO:flaml.tune.searcher.blendsearch:No low-cost partial config given to the search algorithm. For cost-frugal search, consider providing low-cost values for cost-related hps via 'low_cost_partial_config'. More info can be found at https://microsoft.github.io/FLAML/docs/FAQ#about-low_cost_partial_config-in-tune


Busca finalizada!


In [5]:
print("Melhor modelo encontrado:", automl.best_estimator)
print("Acurácia estimada na validação interna:", 1 - automl.best_loss)

Melhor modelo encontrado: extra_tree
Acurácia estimada na validação interna: 0.9962264150943396


In [6]:
from sklearn.metrics import classification_report, accuracy_score, confusion_matrix
import matplotlib.pyplot as plt

y_pred = automl.predict(X_test)
test_acc = accuracy_score(y_test, y_pred)

print(f"Acurácia no conjunto de teste: {test_acc:.4f}")
print("-" * 60)
print(classification_report(y_test, y_pred))

Acurácia no conjunto de teste: 0.9851
------------------------------------------------------------
              precision    recall  f1-score   support

      Adelie       1.00      0.97      0.98        29
   Chinstrap       0.93      1.00      0.97        14
      Gentoo       1.00      1.00      1.00        24

    accuracy                           0.99        67
   macro avg       0.98      0.99      0.98        67
weighted avg       0.99      0.99      0.99        67



---

## 6. Discussão e Considerações Finais

### 6.1 Os resultados preliminares são satisfatórios?
Sim. O AutoML alcançou uma acurácia de teste próxima de **98-100%** com apenas 60 segundos de busca, demonstrando que as características morfológicas do dataset *Palmer Penguins* são preditores quase perfeitos para a separação das três espécies.

### 6.2 Comparação com a Estimativa Básica (Moda)
A classe majoritária no dataset é a espécie *Adelie*, representando cerca de 44% das observações.
*   **Baseline Trivial (Moda):** Um preditor simples que previsse sempre a classe *Adelie* obteve acurácia de apenas **44%**.
*   **Ganho do Modelo:** O AutoML atinge **>98%** de acurácia, gerando um ganho expressivo de mais de **54 pontos percentuais** sobre o baseline ingênuo.

### 6.3 Adequação do Problema para o Curso
O dataset é perfeito para o andamento do curso. Ele permitirá explorar visualizações bivariadas elegantes no TP2 (como gráficos de dispersão entre tamanho de nadadeira e bico) e aplicar redes neurais e modelos lineares manuais no TP3.
